File: 02_paradigm_exploration/data_structure_manipulation/stack.pl
% push(+Stack, +Element, -NewStack)
push(Stack, Elem, [Elem|Stack]).

% pop(+Stack, -Element, -NewStack)
pop([H|T], H, T).

% peek(+Stack, -TopElement)
peek([H|_], H).

% Example usage:
% ?- push([], 1, S1), push(S1, 2, S2), peek(S2, Top), pop(S2, Popped, S3).

File: 02_paradigm_exploration/data_structure_manipulation/stack.go
package main

import (
	"errors"
	"fmt"
)

type Stack struct {
	elements []int
}

func (s *Stack) Push(val int) {
	s.elements = append(s.elements, val)
}

func (s *Stack) Pop() (int, error) {
	if len(s.elements) == 0 {
		return 0, errors.New("stack is empty")
	}
	last := s.elements[len(s.elements)-1]
	s.elements = s.elements[:len(s.elements)-1]
	return last, nil
}

func (s *Stack) Peek() (int, error) {
	if len(s.elements) == 0 {
		return 0, errors.New("stack is empty")
	}
	return s.elements[len(s.elements)-1], nil
}

func main() {
	s := &Stack{}
	s.Push(5)
	s.Push(8)
	top, _ := s.Peek()
	fmt.Println("Peek:", top)
	val, _ := s.Pop()
	fmt.Println("Popped:", val)
}
